substrate/frame/support/src/traits/tokens/asset_ops.rs

```diff
@@ -0,0 +1,625 @@
+/// The common destroy strategies are:
+/// * [`JustDo`](common_strategies::JustDo)
+/// * [`IfOwnedBy`](common_strategies::IfOwnedBy)
+/// * [`WithWitness`](common_strategies::WithWitness)
+/// * [`IfOwnedByWithWitness`](common_strategies::IfOwnedByWithWitness)
+pub trait DestroyStrategy {
+	/// This type represents a successful asset destruction.
+	/// It will be in the [`Result`] type of the [`Destroy::destroy`] function.
+	type Success;
+}
+
+/// A trait representing the ability of a certain asset to be destroyed.
+///
+/// This trait can be implemented multiple times using different
+/// [`destroy strategies`](DestroyStrategy).
+///
+/// A destroy strategy defines destroy parameters and the result value type.
+pub trait Destroy<Strategy: DestroyStrategy>: AssetDefinition {
+	/// Destroy the asset identified by the given `id` using the provided `strategy`.
+	///
+	/// The ID type is retrieved from the [`AssetDefinition`].
+	fn destroy(id: &Self::Id, strategy: Strategy) -> Result<Strategy::Success, DispatchError>;
+}
+
+/// A strategy for use in the [`Stash`] implementations.
+///
+/// The common stash strategies are:
+/// * [`JustDo`](common_strategies::JustDo)
+/// * [`IfOwnedBy`](common_strategies::IfOwnedBy)
+pub trait StashStrategy {
+	/// This type represents a successful asset stashing.
+	/// It will be in the [`Result`] type of the [`Stash::stash`] function.
+	type Success;
+}
+
+/// A trait representing the ability of a certain asset to be stashed.
+///
+/// This trait can be implemented multiple times using different
+/// [`stash strategies`](StashStrategy).
+///
+/// A stash strategy defines stash parameters.
+pub trait Stash<Strategy: StashStrategy>: AssetDefinition {
+	/// Stash the asset identified by the given `id` using the provided `strategy`.
+	///
+	/// The ID type is retrieved from the [`AssetDefinition`].
+	fn stash(id: &Self::Id, strategy: Strategy) -> Result<Strategy::Success, DispatchError>;
+}
+
+/// A strategy for use in the [`Restore`] implementations.
+/// The common restore strategies are:
+/// * [`JustDo`](common_strategies::JustDo)
+/// * [`IfRestorable`](common_strategies::IfRestorable)
+pub trait RestoreStrategy {
+	/// This type represents a successful asset restoration.
+	/// It will be in the [`Result`] type of the [`Restore::restore`] function.
+	type Success;
+}
+
+/// A trait representing the ability of a certain asset to be restored.
+///
+/// This trait can be implemented multiple times using different
+/// [`restore strategies`](RestoreStrategy).
+///
+/// A restore strategy defines restore parameters.
+pub trait Restore<Strategy: RestoreStrategy>: AssetDefinition {
+	/// Restore the asset identified by the given `id` using the provided `strategy`.
+	///
+	/// The ID type is retrieved from the [`AssetDefinition`].
+	fn restore(id: &Self::Id, strategy: Strategy) -> Result<Strategy::Success, DispatchError>;
+}
+
+/// This modules contains the common asset ops strategies.
+pub mod common_strategies {
```

Comment:
Might be better to move it to its own file

Reply:
Done